### PR TITLE
[release/0.4][DSA] Vendor the SM100-family v2 indexer backward and use it where accepted

### PR DESCRIPTION
**Pull request by ForFishes**
### PR Category

Operator Mechanism

### PR Types

Improvements

### Description

Two changes that only make sense together.

**Vendor cuDNN `3e27c995`** (`packages/paddlefleet_ops/third_party/cudnn-frontend`,
`6956336f` -> `3e27c995`, PFCCLab `paddle/v1.28.0`). The only delta is
`PFCCLab/cudnn-frontend#14`, which widens the two capability gates on the v2
sparse indexer backward from an exact `(10, 0)` to the SM100 family
(`capability[0] == 10`). The narrow gates refused SM103 parts even though the DSA
compile path already targets `sm_103a` (`utils/compiler.py::_ARCH_MAP`) and the
kernel runs there correctly, and sibling code in that repository already treats
SM103 as SM100-family (`sdpa/fwd/api_dsl.py`'s `_allowed_cc`,
`linear_attention/frost/engine.py`'s "SM100-SM103 or SM107"). Nothing else in the
submodule moves: the two pinned trees differ only in those gates and the test
helper that skipped the v2 suite outside `(10, 0)`.

**Reach the backend from PaddleFleet.** cuDNN 1.28's second GEMM stage for this
kernel keeps kernel 1 (the score-grad precompute) verbatim and replaces the GEMM:
`weights` are upcast to fp32 exactly, the per-slot gradient matrix
`A = grad_signal * weights` is split into a two-term bf16 expansion before the
MMA, and `d_weights` is reduced deterministically in-CTA. `csa_indexer_bwd` never
passed `backend=`, so it was unreachable. It is strictly more accurate, so there
is no switch: the backend is requested whenever the vendored cuDNN accepts it.

Measured on an SM103 part, `S=8192 H=64 D=128 topk=2048`, relative L2 against an
fp64 evaluation of the kernel's documented contract -- taken from the grad_signal
the generic backend leaves in `attn_score`, so the shared kernel 1 is factored out
and only the GEMM stage is judged:

| output | generic | v2 |
|---|---|---|
| `d_index_q` (bf16) | 2.321e-03 | 1.659e-03 |
| `d_weights` | 1.655e-03 (bf16) | 1.697e-07 (fp32) |
| `d_index_k` (fp32) | 1.517e-03 | 1.207e-05 |

The `d_weights` row is why the selected path also hands cuDNN an fp32
`d_weights` buffer. The accumulator is fp32 on both backends, so the bf16 buffer
the wrapper allocates by default rounds the entire gain back to the bf16
representation floor: with it, `d_weights` measured 1.655e-03, i.e. identical to
the generic backend. fp32 is only accepted on this backend, and the trailing
`grad_weights.cast` restores the caller's dtype, so the buffer dtype stays an
internal detail.

The cost is speed, and only at the top of the topk range: the two-term expansion
doubles the GEMM work, so it wins where the GEMM does not dominate and loses
where it does. Swept on an idle SM103 part at the shape above, min of 3
alternating rounds of 20 iterations, spread within a round under 1.5%:

| topk | generic | v2 | speedup |
|---|---|---|---|
| 128 | 0.457 ms | 0.293 ms | 1.559x |
| 256 | 0.643 ms | 0.537 ms | 1.197x |
| 512 | 0.909 ms | 0.879 ms | 1.035x |
| 1024 | 1.550 ms | 1.579 ms | 0.982x |
| 2048 | 2.910 ms | 2.984 ms | 0.975x |

The crossover sits between 512 and 1024, so a `topk=2048` configuration pays
about 2.5% on this one kernel. That is deliberately not gated on topk: two orders
of magnitude on `d_index_k` is worth 2.5% of one backward kernel, and a
topk-dependent backend would make the numerics a function of the sequence budget,
which is harder to reason about than a uniform 2.5%.

`_select_backend` mirrors the vendored cuDNN's envelope -- the `backend`
parameter must exist on the installed wrapper, capability major 10, `H == 64`,
`D == 128`, `block_I == 128`, `topk % 128 == 0` with `128 <= topk <= 2048` -- and
is `lru_cache`d on the shape because it would otherwise re-run on every indexer
backward of every layer. Requesting outside the envelope aborts the step rather
than selecting a slower kernel, which is why the predicate is evaluated up front
instead of caught.

One case cannot be settled up front: **an installed `paddlefleet_ops` older than
this checkout's submodule**, which is the normal state between a bump and a
rebuild. Its cuDNN still has the narrow gate, so it refuses an SM103 device that
the predicate above accepts. That refusal is caught -- narrowly. cuDNN has no
dedicated exception type, so the two refusal messages are matched by substring;
both fire strictly before `execute` (`check_support` at plan construction, and the
kernel factory from `compile`, api.py:475 inside `compile` at :457 while `execute`
starts at :534), so the score buffers are untouched and the retry on the generic
backend is exact rather than approximate. Everything else propagates: catching
`RuntimeError` broadly would also swallow a compile or launch failure raised
*after* kernel 1 had overwritten `attn_score` with the grad signal, and the retry
would then compute gradients from a consumed buffer -- silently wrong, with the
real error permanently downgraded. The refusal is latched so it costs one plan
construction per process, and logged once.

Two consequences worth stating, since neither is visible from the diff:

* **The numerical trajectory changes** wherever the backend is selected. No
  checkpoint format, parameter or `state_dict` key changes -- this is a
  gradient-precision change on the indexer's own five weights per layer, and the
  backbone receives `grad_output` untouched -- but loss curves are no longer
  bit-comparable across this commit.
* **`d_index_k` stays non-deterministic** on both backends (fp32 atomics either
  way). The v2 determinism claim covers `d_weights` and `d_index_q` only, so
  "deterministic v2" must not be read as "bitwise reproducible step".

Tests: `tests/single_card_tests/custom_ops/test_cudnn_dsa_indexer_bwd_backend.py`
(15 tests / 26 subtests, host-side, wrapper mocked, no cuDNN build needed) covers
every envelope condition on both sides of its boundary -- including the device
family, so SM100/SM103/SM107 select and SM90/SM120 do not, and a wrapper that
predates the `backend` parameter -- plus the argument contract (accepted calls
carry `backend=` and an fp32 `d_weights`; rejected calls are argument-for-argument
what they were before; the caller's dtypes come back unchanged) and the fallback:
both refusal messages fall back with the score buffers verified byte-identical to
the first attempt, the refusal latches so the next step does not re-attempt, and
three non-refusal errors (illegal memory access, compile failure, a `ValueError`)
propagate instead of being downgraded.

Existing suites re-run green on an SM103 part against a version-matched tree:
`test_cudnn_dsa_indexer_bwd` 35, `test_cudnn_dsa_indexer_fwd` 29,
`test_cudnn_dsa_indexer_topk_column_mask` 9 + 120 subtests,
`test_cudnn_dsa_indexer_shape_guard` 8, `test_indexer_loss_overlap` 64 + 22
subtests.

Same change on `develop`: #1966 (the `Validate Dev PR` gate stays red until that one merges).

### CODE REVIEW

**Comment by ForFishes** [comment]
P1 已修，谢谢——这条是对的，我原来的 `except (RuntimeError, ValueError)` 确实过宽。

**问题**：那个宽捕获会把 kernel 1 已经用 grad signal 覆写过 `attn_score` **之后**才抛出的编译/执行失败也吞掉，重试 generic 就会拿一份已被消费的 buffer 算梯度——静默错误；同时把真实执行错误永久降级成一次 fallback。

**改法**：只匹配 cuDNN 为这个 backend 抛出的两条**设备拒绝**消息，其余一律 `raise`：

```python
_V2_REFUSAL_MARKERS = (
    "backend='sm100_v2' requires an SM100",
    "indexer_backward_v2_sm100 requires SM100",
)

def _is_v2_device_refusal(exc):
    return isinstance(exc, RuntimeError) and any(
        m in str(exc) for m in _V2_REFUSAL_MARKERS
    )
...
        except RuntimeError as exc:
            if not _is_v2_device_refusal(exc):
                raise
```

用消息匹配是因为 cuDNN 没有为这两种拒绝定义专用异常类型。**这两条都严格发生在 `execute` 之前**，所以 buffer 一定未被改写：

- `check_support` 在建 plan 阶段抛（`api.py:975` 的 `assert obj.check_support()`，早于 `:979` 的 `obj.execute()`）
- kernel 工厂从 `compile` 里抛（`api.py:475` 在 `compile` 内，`compile` 起于 `:457`，而 `execute` 起于 `:534`）

也就是说 kernel 1 只在 `execute` 里消费 score buffer，两条拒绝都在那之前，重试是**精确**的而不是近似的。

**为什么还需要这个 fallback**：本 PR 同时把 vendored cuDNN bump 到 `3e27c995`（含 `PFCCLab/cudnn-frontend#14`，把设备门从精确 `(10, 0)` 放宽到 SM100 家族）。但**装着的 `paddlefleet_ops` wheel 可能比 checkout 里的 submodule 旧**——bump 之后、重编之前就是这个状态——那时它自带的 cuDNN 门更窄，会拒绝一个上面判据接受的 SM103 设备。这种情况回退比中断训练更合理，梯度仍然正确，只是精度低一档。

新增三个测试专门守这条：两条拒绝消息都会回退（并逐位断言重试收到的 score buffer 与首次尝试完全相同），拒绝会被 latch 住（下一步不再重试），以及**三种非拒绝异常（illegal memory access / 编译失败 / `ValueError`）必须原样抛出、且不会置上 latch**。

单测：新文件 15 passed + 26 subtests；回归 `test_cudnn_dsa_indexer_bwd` 35、`_fwd` 29、`_topk_column_mask` 9+120、`_shape_guard` 8、`test_indexer_loss_overlap` 64+22，全绿。

⚠️ 顺带说明：这一版把 `packages/paddlefleet_ops/third_party/cudnn-frontend` 也动了（`6956336f` → `3e27c995`），所以会重新触发 `packages/` 的 approve 要求。放宽判据和 vendor 更宽的 cuDNN 必须在同一个 commit 里，否则任一方单独落地都是错的：先放宽会在旧 cuDNN 上崩，先 bump 则判据仍然拦着、拿不到收益。

**Review by sneaxiy — APPROVED** [review]
